File: delta/groups.ts
import { Composer, Context, InlineKeyboard } from "../deps.ts";
import communities from "../communities.json" assert { type: "json" };
import pager from "../utils/pager.ts";

const composer = new Composer();
const ctxMenuText =
  "<b>Telegramdagi Rust Hamjamiyatlari yoki Guruhlari:</b>\nAgar o'zingizni guruhingizni qo'shmoqchi bo'lsangiz, bizni <a href='https://github.com/rust-lang-uz/telegram/blob/main/communities.json'>community.json</a> ni yangilang!";

composer.command("group", async (ctx: Context): Promise<void> => {
  const keyboard = new InlineKeyboard();

  for (const community of pager(1)) {
    keyboard.text(
      community.name,
      `detail_${1}_${community.telegram.replace("@", "")}`,
    ).row();
  }

  if (pager(2).length > 0) {
    keyboard.text(`Keyingi ➡️`, `group_2`);
  }

  await ctx.reply(ctxMenuText, {
    parse_mode: "HTML",
    reply_markup: keyboard,
    disable_web_page_preview: true,
  });
});

composer.callbackQuery(/^group_(\d+)$/, async (ctx: Context) => {
  const page = Number(ctx.match![1]);
  const keyboard = new InlineKeyboard();

  for (const community of pager(page)) {
    keyboard.text(
      community.name,
      `detail_${page}_${community.telegram.replace("@", "")}`,
    ).row();
  }

  if (pager(page - 1).length > 0) {
    keyboard.text(`⬅️ Oldingi`, `group_${page - 1}`);
  }

  if (pager(page + 1).length > 0) {
    keyboard.text(`Keyingi ➡️`, `group_${page + 1}`);
  }

  await ctx.editMessageText(ctxMenuText, {
    parse_mode: "HTML",
    reply_markup: keyboard,
    disable_web_page_preview: true,
  });
});

composer.callbackQuery(/^detail_(\d+)_(.*)$/, async (ctx: Context) => {
  const keyboard = new InlineKeyboard();
  const page = ctx.match![1];
  const result = communities.filter((com) =>
    com.telegram.replace("@", "") === ctx.match![2]
  );

  if (result.length) {
    const data = result[0];

    if (data.telegram) {
      keyboard.url(
        `Telegram`,
        `https://t.me/${data.telegram.replace("@", "")}`,
      );
    }

    if (data.link) {
      keyboard.url(`Web`, data.link);
    }

    keyboard.row().text(`🔙 Orqaga`, `group_${page}`);

    await ctx.editMessageText(
      `<b>${data.name}</b>` +
      `\n` +
      `\n` +
      `<i>${data.about}</i>` +
      `\n` +
      `\n` +
      `<b>Use the following buttons to get to the links:</b>`,
      {
        parse_mode: "HTML",
        reply_markup: keyboard,
      },
    );
  } else {
    await ctx.editMessageText(`<b>Ushbu guruh mavjud emas!</b>`, {
      parse_mode: "HTML",
      reply_markup: new InlineKeyboard().text(`🔙 Orqaga`, `group_${page}`),
    });
  }
});

export default composer;

File: src/functions/groups.rs
use crate::utils::{
    groups::{Group, Groups},
    keyboard::Keyboard,
    message::Rustina,
};
use teloxide::{
    payloads::{EditMessageTextSetters, SendMessageSetters},
    prelude::*,
    types::{InlineKeyboardMarkup, ParseMode},
};

static TEXT: &str = "<b>Telegramdagi Rust Hamjamiyatlari yoki Guruhlari:</b>\nAgar o'zingizni guruhingizni qo'shmoqchi bo'lsangiz, bizni <a href='https://github.com/rust-lang-uz/rustina/blob/main/data/communities.json'>community.json</a> ni yangilang!";

pub async fn command(bot: &Bot, msg: &Message, groups: &Groups) -> ResponseResult<()> {
    bot.send_message_tf(msg.chat.id, TEXT, msg)
        .parse_mode(ParseMode::Html)
        .reply_markup(keyboard_list(groups, 1))
        .disable_web_page_preview(true)
        .await?;

    Ok(())
}

pub async fn callback_list(
    bot: &Bot,
    q: &CallbackQuery,
    args: &[&str],
    groups: &Groups,
) -> ResponseResult<()> {
    if !args.is_empty() {
        if let Some(Message { id, chat, .. }) = q.message.clone() {
            bot.edit_message_text(chat.id, id, TEXT)
                .parse_mode(ParseMode::Html)
                .reply_markup(keyboard_list(groups, args[0].parse().unwrap_or(1)))
                .disable_web_page_preview(true)
                .await?;
        } else if let Some(id) = q.inline_message_id.clone() {
            bot.edit_message_text_inline(id, "Oopsie, something went wrong...")
                .await?;
        }
    }

    Ok(())
}

pub async fn callback_detail(bot: &Bot, q: &CallbackQuery, args: &[&str]) -> ResponseResult<()> {
    let groups: Groups = Groups::new();
    let find = groups.find_group(args[1..].join("_").to_string());

    if !args.is_empty() {
        if let Some(Message { id, chat, .. }) = q.message.clone() {
            bot.edit_message_text(chat.id, id, view_detail(&find))
                .parse_mode(ParseMode::Html)
                .reply_markup(keyboard_detail(args[0].parse().unwrap_or(1), &find))
                .await?;
        } else if let Some(id) = q.inline_message_id.clone() {
            bot.edit_message_text_inline(id, "Oopsie, something went wrong...")
                .await?;
        }
    }

    Ok(())
}

pub fn view_detail(data: &Option<Group>) -> String {
    match data {
        Some(d) => {
            format!(
                "<b>{}</b>\n\n<i>{}</i>\n\n<b>Use the following buttons to get to the links:</b>",
                d.name, d.about
            )
        }
        None => "<b>Ushbu guruh mavjud emas!</b>".to_string(),
    }
}

pub fn keyboard_list(groups: &Groups, page: i32) -> InlineKeyboardMarkup {
    let mut keyboard = Keyboard::new();

    for group in groups.get_groups(page, 5) {
        keyboard.text(
            &group.name,
            &format!(
                "detail_{}_{}",
                page,
                group.telegram.clone().replace('@', "")
            ),
        );
        keyboard.row();
    }

    if !groups.get_groups(page + 1, 5).is_empty() {
        keyboard.text("Keyingi ➡️", &format!("group_{}", page + 1));
    }

    if page > 1 {
        keyboard.text("⬅️ Oldingi", &format!("group_{}", page - 1));
    }

    keyboard.get()
}

pub fn keyboard_detail(page: i32, data: &Option<Group>) -> InlineKeyboardMarkup {
    let mut keyboard = Keyboard::new();

    if let Some(group) = data {
        keyboard.url("Telegram", &format!("https://t.me/{}", group.telegram));

        if group.link.is_some() {
            keyboard.url("Web", &group.link.clone().unwrap());
        }

        keyboard.row();

        keyboard.text("🔙 Orqaga", &format!("group_{}", page));
    } else {
        keyboard.text("🔙 Orqaga", &format!("group_{}", page));
    }

    keyboard.get()
}
